File: 1028. Recover a Tree From Preorder Traversal/neet.go
package main

import "strconv"

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func recoverFromPreorder(traversal string) *TreeNode {
	stack := make([]*TreeNode, 0)
	dashes, i := 0, 0

	for i < len(traversal) {
		if traversal[i] == '-' {
			dashes++
			i++
			continue
		}
		j := i
		for j < len(traversal) && traversal[j] != '-' {
			j++
		}
		val, _ := strconv.Atoi(traversal[i:j])
		node := &TreeNode{Val: val}

		for len(stack) > dashes {
			stack = stack[:len(stack)-1]
		}

		if len(stack) > 0 && stack[len(stack)-1].Left == nil {
			stack[len(stack)-1].Left = node
		} else if len(stack) > 0 {
			stack[len(stack)-1].Right = node
		}

		stack = append(stack, node)
		dashes = 0
		i = j
	}

	return stack[0]
}

File: 1028. Recover a Tree From Preorder Traversal/neet.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {string} traversal
 * @return {TreeNode}
 */
var recoverFromPreorder = function (traversal) {
  const stack = [];
  let dashes = 0;
  let i = 0;
  while (i < traversal.length) {
    if (traversal[i] === "-") {
      dashes++;
      i++;
      continue;
    }
    let j = i;
    while (j < traversal.length && traversal[j] !== "-") j++;
    const val = Number(traversal.slice(i, j));
    const node = new TreeNode(val);
    while (stack.length > dashes) stack.pop();
    if (stack.length > 0 && stack[stack.length - 1].left === null) {
      stack[stack.length - 1].left = node;
    } else if (stack.length > 0) {
      stack[stack.length - 1].right = node;
    }
    stack.push(node);
    dashes = 0;
    i = j;
  }
  return stack[0];
};
